lib/auth/auth_with_roles.go

```diff
@@ -1193,31 +1208,22 @@ func (a *ServerWithRoles) generateUserCerts(ctx context.Context, req proto.UserC
 		if req.Expires.Before(a.authServer.GetClock().Now()) {
 			return nil, trace.AccessDenied("access denied: client credentials have expired, please relogin.")
 		}
-		// If the user is generating a certificate, the roles and traits come from
-		// the logged in identity.
+	}
+
+	// If the user is generating a certificate, the roles and traits come from the logged in identity.
+	if req.Username == a.context.User.GetName() {
 		roles, traits, err = services.ExtractFromIdentity(a.authServer, a.context.Identity.GetIdentity())
 		if err != nil {
 			return nil, trace.Wrap(err)
 		}
-	default:
-		err := trace.AccessDenied("user %q has requested to generate certs for %q.", a.context.User.GetName(), req.Username)
-		log.Warning(err)
-		if err := a.authServer.emitter.EmitAuditEvent(a.CloseContext(), &events.UserLogin{
-			Metadata: events.Metadata{
-				Type: events.UserLoginEvent,
-				Code: events.UserLocalLoginFailureCode,
-			},
-			Method: events.LoginMethodClientCert,
-			Status: events.Status{
-				Success:     false,
-				Error:       trace.Unwrap(err).Error(),
-				UserMessage: err.Error(),
-			},
-		}); err != nil {
-			log.WithError(err).Warn("Failed to emit local login failure event.")
+	} else {
+		// Do not allow combining impersonation and access requests
+		if len(req.AccessRequests) > 0 {
+			log.WithError(err).Warningf("User %v tried to issue a cert for %v and added access requests. This is not supported.", a.context.User.GetName(), req.Username)
+			return nil, trace.AccessDenied("access denied")
 		}
```

Comment:
Per my comment above, impersonated certs need to be able to perform normal reissues in order to set things like route-to-cluster directives.  To preserve this check's current behavior after that change, it should be moved to the outer scope and changed to something like this:

```golang
if a.context.Identity != nil && a.context.Identity.GetIdentity().Impersonator != "" && len(req.AccessRequests) > 0 {
    // ...
}
```

Also, note: When integrating this with the upcoming access review feature, we should prevent users from either reviewing their own requests under an assumed identity, or double-reviewing requests via assumed identities.